src/lib/blit.dasl

```diff
@@ -0,0 +1,138 @@
+-- blit.lua - offload engine for memory operations
+
+module(..., package.seeall)
+
+local dasm = require("dasm")
+local ffi = require("ffi")
+
+|.arch x64
+|.actionlist actions
+
+-- The blit module provides "blitter" operation to offload
+-- performance-critical memory operations. The API allows scheduling a
+-- series of operations, that can be performed at any time and in any
+-- order, and then executing a "barrier" to wait for completion.
+
+-- This module implements a blitter that defers all copy operations
+-- until the barrier() is invoked and then executes them with a single
+-- assembler code routine.
+
+ffi.cdef[[
+  struct blit_queue_entry {
+    void *src;
+    void *dst;
+    uint64_t len;
+  };
+]]
+
+-- Queue of memory operations
+local maxqueue = 10000
+local queue = ffi.new("struct blit_queue_entry[?]", maxqueue+2)
+local nqueued = 0
+
+function copy (dst, src, len)
+   -- XXX This routine is hard-coded for multiples of 32 bytes.
+   assert(len%32 == 0)
+   if nqueued == maxqueue then barrier() end
+   queue[nqueued].src = src
+   queue[nqueued].dst = dst
+   queue[nqueued].len = len
+   nqueued = nqueued + 1
+   -- Sentinel
+   queue[nqueued].src = nil
+end
+
+-- Assembler code for the barrier operation
+function gen_barrier (Dst)
+   | mov64 r8, queue
+
+   |->queue:
+   | cmp qword [r8], 0          -- sentinel?
+   | je >9
+
+   -- Load parameters for the next copy
+   | mov rsi, [r8]     -- source
+   | mov rdi, [r8+8]   -- destination
+   | mov rcx, [r8+16]  -- length
+
+   -- Copy 32 bytes at a time
+   | xor rax, rax
+   |->copy:
+   | vmovdqu ymm0, [rsi+rax]
+   | vmovdqu [rdi+rax], ymm0
```

Comment:
Yes, delicate indeed :)  One thing to try is instead of doing load, store, load, store, to do load, load, store, store.  That was what worked best for me.  Good luck :)